pkg/controller/jobs/leaderworkerset/leaderworkerset_reconciler.go

```diff
@@ -92,36 +99,69 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	log := ctrl.LoggerFrom(ctx)
 	log.V(2).Info("Reconcile LeaderWorkerSet")
 
-	err = r.createPrebuiltWorkloadsIfNotExist(ctx, lws)
+	wlList := &kueue.WorkloadList{}
+	if err := r.client.List(ctx, wlList, client.InNamespace(lws.GetNamespace()),
+		client.MatchingFields{indexer.OwnerReferenceUID: string(lws.GetUID())},
+	); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	toCreate, toFinalize := r.matchWorkloads(lws, wlList.Items)
+
+	eg, ctx := errgroup.WithContext(ctx)
+
+	eg.Go(func() error {
+		return parallelize.Until(ctx, len(toCreate), func(i int) error {
+			return r.createPrebuiltWorkload(ctx, lws, toCreate[i])
+		})
+	})
+
+	eg.Go(func() error {
+		return parallelize.Until(ctx, len(toFinalize), func(i int) error {
+			return r.removeOwnerReference(ctx, lws, toFinalize[i])
+		})
+	})
+
+	err = eg.Wait()
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
 	return ctrl.Result{}, nil
 }
 
-func (r *Reconciler) createPrebuiltWorkloadsIfNotExist(ctx context.Context, lws *leaderworkersetv1.LeaderWorkerSet) error {
-	replicas := ptr.Deref(lws.Spec.Replicas, 1)
+// matchWorkloads matches existing workloads against the desired state in LeaderWorkerSet,
+// determining which workloads need to be created and which should be finalized.
+//
+// It takes a LeaderWorkerSet and a slice of existing Workload objects as input and returns:
+// 1. A slice of workload names that need to be created
+// 2. A slice of Workload pointers that need to be finalized
+func (r *Reconciler) matchWorkloads(lws *leaderworkersetv1.LeaderWorkerSet, existingWorkloads []kueue.Workload) ([]string, []*kueue.Workload) {
+	var (
+		toCreate   []string
+		toFinalize = utilslices.ToRefMap(existingWorkloads, func(e *kueue.Workload) string {
+			return e.Name
+		})
+		replicas = ptr.Deref(lws.Spec.Replicas, 1)
+	)
+
 	for i := int32(0); i < replicas; i++ {
-		if err := r.createPrebuiltWorkloadIfNotExist(ctx, lws, i); err != nil {
-			return err
+		workloadName := GetWorkloadName(lws.UID, lws.Name, fmt.Sprint(i))
+		if _, ok := toFinalize[workloadName]; ok {
```

Comment:
Is this correct? Why do we want to delete from the map if it's there? Did you mean 
```_, ok := toFinalize[workloadName]; !ok```?

Reply:
Yes, it's correct. We create a map with all workloads and remove the ones we want to keep. As a result, the remaining workloads in the map are the ones we need to delete.

Reply:
I see, makes sense, thanks